pkg/query/columnquery.go

```diff
@@ -274,29 +274,44 @@ func (q *ColumnQueryAPI) Query(ctx context.Context, req *pb.QueryRequest) (*pb.Q
 	}()
 
 	if req.GetReportType() == pb.QueryRequest_REPORT_TYPE_PROFILE_METADATA {
-		mappingFiles, err := q.getMappingFiles(ctx, req.GetMerge())
+		var profileMetadata *pb.ProfileMetadata
 
-		labels, labels_err := q.getLabels(ctx, &pb.LabelsRequest{
-			Match: []string{},
-			Start: req.GetMerge().Start,
-			End:   req.GetMerge().End,
-		})
+		switch req.Mode {
+		case pb.QueryRequest_MODE_MERGE:
+			mappingFiles, labels, err := getMappingFilesAndLabels(ctx, q.querier, req.GetMerge().Query, req.GetMerge().Start.AsTime(), req.GetMerge().End.AsTime())
+			if err != nil {
+				return nil, err
+			}
 
-		if err != nil {
-			return nil, fmt.Errorf("getting mapping files: %w", err)
-		}
 
-		if labels_err != nil {
-			return nil, fmt.Errorf("getting labels: %w", labels_err)
+			profileMetadata = &pb.ProfileMetadata{
+				MappingFiles: mappingFiles,
+				Labels:       labels,
+			}
+
+		case pb.QueryRequest_MODE_DIFF:
+			mappingFiles_a, labels_a, err := getMappingFilesAndLabels(ctx, q.querier, req.GetDiff().A.GetMerge().GetQuery(), req.GetDiff().A.GetMerge().Start.AsTime(), req.GetDiff().A.GetMerge().End.AsTime())
+			if err != nil {
+				return nil, err
+			}
+
+			mappingFiles_b, labels_b, err := getMappingFilesAndLabels(ctx, q.querier, req.GetDiff().B.GetMerge().GetQuery(), req.GetDiff().B.GetMerge().Start.AsTime(), req.GetDiff().B.GetMerge().End.AsTime())
+			if err != nil {
+				return nil, err
+			}
+
+			profileMetadata = &pb.ProfileMetadata{
+				MappingFiles: append(mappingFiles_a, mappingFiles_b...),
```

Comment:
We can’t just append them, we need to merge them. But good news is that they’re both sorted so something like a k-way merge is possible.